File: MonotonicStack/2552-maximum-sum-of-distinct-subarrays-with-length-k/maximum-sum-of-distinct-subarrays-with-length-k.py
class Solution:
    def maximumSubarraySum(self, nums: List[int], k: int) -> int:

        n=len(nums)
        l=0
        vis=set()
        maximum=0
        sumi=0
        for r in range(n):
            
            while nums[r] in vis:
                vis.remove(nums[l])
                sumi-=nums[l]
                l+=1
            vis.add(nums[r])
            sumi+=nums[r]
            # print(f'{l}{r}{vis}')
            if (r-l+1)==k:
                maximum=max(maximum,sumi)
                sumi-=nums[l]
                vis.remove(nums[l])
                l+=1
            # print(f'{l}{r}{vis}')
            # print(f'{vis} {nums[r]} {l}')

        
        return maximum

            

            

        

        

File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/maximum-sum-of-distinct-subarrays-with-length-k.java
class Solution {
    public long maximumSubarraySum(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        long result = 0;
        int l = 0;
        long sum=0;
        for (int r = 0; r < nums.length; r++) {
            map.merge(nums[r], 1, Integer::sum);
            sum+=nums[r];
           while (map.get(nums[r]) > 1) {
               map.replace(nums[l], map.get(nums[l]) - 1);
               sum-=nums[l];
               if(map.get(nums[l])==0)
               {
                   map.remove(nums[l]);
               }
               l++;
           }
            if ((r-l+1)== k) {
                result=Math.max(result,sum);
                map.replace(nums[l], map.get(nums[l]) - 1);
                if(map.get(nums[l])==0)
                {
                    map.remove(nums[l]);
                }
                sum-=nums[l];
                l++;
            }

        }
        return result;

        
    }
}